### PR TITLE
Fix binding <select> to a null value

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This comes out of investigating https://github.com/dotnet/aspnetcore/issues/13440. The original issue, #13440, no longer repros in 3.1.x. However some of the later comments on the issue are reporting a different problem that does still repro.

Specifically, how do you have a `<select>` and bind it to a property whose value is sometimes null, for example a nullable-enum value?

This is unfortunately quite a problematic scenario because of a historical quirk in HTML. For any `<option>` that has no `value` attribute, the value is considered to be the *text content* of the option. This is a huge mess because (1) it takes away the only chance we had for representing the idea of a null value (not an empty string), and (2) it leads to completely unpredictable behavior because the text content might or might not happen to match up with a known server-side value depending on the user's current locale.

So since we have no way to have a true null-valued `<option>`, the closest we can do is to forget about distinguishing null from empty-string, and ensure that empty-string can round-trip in a binding and is treated as null on the .NET side, at least in the case of a nullable enum.

As it happens, our binding code that parses an incoming value as a nullable enum *already* treats `string.Empty` as `null`, so all that remains is accepting a rendered `null` value (represented by the removal/absence of the parent `<select>`'s `value` attribute) as mapping back to the empty-string `<option>`.

I'm not happy with the fact that we can't really have proper null values, but without either (1) going back to 1995 or (2) piling hacks upon hacks to fight against the browser's default behavior, this is the closest I think we can get.

### CODE REVIEW

**Comment by pranavkm** [comment]
> going back to 1995

Should we bring that up in sprint planning?

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 1]
```diff
@@ -519,3 +523,15 @@ function stripOnPrefix(attributeName: string) {
...
+function setSelectElementValue(element: HTMLSelectElement, value: string | null) {
+  // There's no sensible way to represent a select option with value 'null', because
+  // (1) HTML attributes can't have null values - the closest equivalent is absence of the attribute
+  // (2) When picking an <option> with no 'value' attribute, the browser treats the value as being the
+  //     *text content* on that <option> element. Trying to suppress that default behavior would involve
+  //     a long chain of special-case hacks, as well as being breaking vs 3.x.
+  // So, the most plausible 'null' equivalent is an empty string. It's unfortunate that people can't
+  // write <option value=@someNullVariable>, and that we can never distinguish between null and empty
+  // string in a bound <select>, but that's a limit in the representational power of HTML.
+  element.value = value || '';
```
Comment (pranavkm) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:536]: Could we document this behavior in https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-3.1? /cc @guardrex

**Review by pranavkm — APPROVED** [review]

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts (hunk 1 above)
Comment (guardrex) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:536]: Is it more general tho than just forms validation? Seems like a *binding topic* subject that we could cross-link to from the forms validation topic.

Comment (guardrex) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:536]: Well ... we already have radio buttons covered there, so I suppose that's fine. I'll look at cross-linking this the other way. I'll cross-link from the binding topic to the forms validation topic the same way that we do for radio button coverage.

**Comment by mrpmorris** [comment]
Would it be possible for JS to set the `option` `value` to `undefined` when we try to set the value to `@null`, and in the Blazor code detect that?

**Comment by guardrex** [comment]
@mrpmorris ... You'll need to open a new issue (or comment on an existing open issue). The team doesn't work off of closed issue or closed PR comments.

**Pull request merged**